asf_public_discourse_home_decarbonisation/utils/preprocessing_utils.py

```diff
@@ -0,0 +1,161 @@
+import pandas as pd
+from nltk.corpus import stopwords
+from nltk.util import bigrams, trigrams
+from collections import Counter
+import re
+from typing import List
+from nltk.probability import FreqDist
+
+
+######### 3. Word cloud of frequently used words in posts ############
+def preprocess_text(dataframe, custom_stopwords) -> List:
+    """
+    Preprocesses text data for word cloud generation.
+
+    Args:
+        dataframe (pandas.DataFrame): The DataFrame containing text data in a column named 'text'.
+        custom_stopwords (list): A list of additional stopwords to be removed from the text.
+
+    Returns:
+        dict: A dictionary of word frequencies filtered based on the defined threshold.
+    """
+    # Combine and lowercase text data
+    text_data = " ".join(dataframe["text"].astype(str)).lower()
+
+    # Update stop words
+    stop_words = set(stopwords.words("english"))
+    stop_words.update(custom_stopwords)
+
+    # Tokenize the text into words and remove stopwords
+    tokens = text_data.split()
+    filtered_tokens = [word for word in tokens if word not in stop_words]
+
+    return filtered_tokens
+
+
+######## 4.Generate bigram and trigram frequency distributions ########
+
+
+def process_ngrams(filtered_tokens, n):
+    """
+    Processes n-grams (bigrams or trigrams) from the given tokens and filters them based on a threshold.
+
+    Args:
+        filtered_tokens (list): A list of tokens from which to form n-grams.
+        n (int): The size of the n-grams (2 for bigrams, 3 for trigrams).
+
+    Returns:
+        tuple: A tuple containing two dictionaries, the raw and filtered n-gram frequency distributions.
+    """
+    if n == 2:
+        raw_ngram_freq_dist = FreqDist(bigrams(filtered_tokens))
+        threshold_multiplier = 0.0002
```

Comment:
I am happy to move these to the analysis file! We discussed this on Thursday, didn't come to a final decision but I guess the need to avoid confusing the reader but also portraying accurate info is the balance.

Reply:
I am coming down on the side of just plotting the top 100 words (an example) and giving an accurate frequency filter (i.e. it may change for different plots).

Reply:
Happy for it to stay like this for now. In general I prefer to do:
- Get all tokens/n-grams with frequency >= X (e.g 100) or above;
- Out of those, plot the top Y (e.g. 30)

If there aren't 30 to plot (because only a few have the required frequency), it will plot as many as there are. If there are more than 30, you're plotting the most common 30. But at least you know they all have frequency above 100 (or what ever one fixes). This way, we know our acceptable threshold is met.